### PR TITLE
Truncate check report to accommodate GitHub limit

### PR DESCRIPTION
**Pull request by tatchley-drizly**
* Added logic to tail the check report, but preserve the JSON if the check summary file was too large
* I didn't handle the edge case when the JSON was larger than GitHub's limit, but can address it if we can define how it should be handled (drop assignments until we are within the limit?)

### CODE REVIEW

**Comment by ikatz-drizly** [comment]
Is `dist/index.js` generated from the other files in this repo?

File: src/inputs.ts [hunk 1]
```diff
@@ -4,6 +4,30 @@ import * as fs from 'fs';
...
+// getBinarySize and truncateToBinarySize were adapted from unlicensed work here:
+// https://github.com/lemonde/utf8-binary-cutter/blob/master/lib/utf8-binary-cutter.js
+function getBinarySize(s: string): number {
+  // Get binary size of a UTF8 string
+  return Buffer.byteLength(s || '', 'utf8');
+}
+
+function truncateToBinarySize(string: string, binaryMaxSize: number): string {
+  // Truncate a UTF8 string to less than or equal to binaryMaxSize
+  const DEFAULT_TRUNCATE_STRING = '...';
```
Comment (ikatz-drizly) [inline src/inputs.ts:16]: It seems like these should live outside the function itself

**Comment by tatchley-drizly** [comment]
> Is `dist/index.js` generated from the other files in this repo?

Correct

File: src/inputs.ts [hunk 2]
```diff
@@ -4,6 +4,30 @@ import * as fs from 'fs';
...
+  const DEFAULT_TRUNCATE_STRING_BINARY_SIZE = Buffer.byteLength(DEFAULT_TRUNCATE_STRING, 'utf8');
+
+  string = string || '';
```
Comment (ikatz-drizly) [inline src/inputs.ts:19]: This is an odd thing to do, it might need a comment.  If it's a coercion to string, you should explain why this is the right one.  If it's a guard against `null` you should consider just `if (string == null) return ''`.

**Comment by tatchley-drizly** [comment]
I'm just noticing this issue...
the log of an [error'd run](https://github.com/DrizlyInc/distributor/runs/3352366420?check_suite_focus=true) specifies 65535 _characters_, rather than 65535 _bytes_ (though I need to change the `65536` constant I put anyway). I think this may call for a reworking on the PR

File: src/inputs.ts [hunk 3]
```diff
@@ -17,12 +41,51 @@ const parseJSON = <T>(getInput: GetInput, property: string): T | undefined => {
...
+function reverseString(input_string: string): string {
+  return input_string.split('').reverse().join('');
+}
+
+function buildOutputDescription(
+  output_description_md: string,
+  json_vars: string,
+  max_allowed_bytes: number,
+): string {
+  // DESCRIPTION:
```
Comment (ikatz-drizly) [inline src/inputs.ts:53]: We should write these as javadoc (parseable) comments:
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns

File: src/inputs.ts [hunk 4]
```diff
@@ -17,12 +41,51 @@ const parseJSON = <T>(getInput: GetInput, property: string): T | undefined => {
...
+  // RETURNS:
+  // a formatted output description file of no more than max_allowed_bytes
+
+  const output_file_size = getBinarySize(output_description_md);
+  const json_output_markdown_size = getBinarySize(json_vars);
+
+  if (output_file_size + json_output_markdown_size <= max_allowed_bytes) {
+    // everything is within bounds, carry on
+    return output_description_md + json_vars;
+  } else {
```
Comment (ikatz-drizly) [inline src/inputs.ts:69]: since the `if` branch returns, the `else { }` is unnecessary here.

it also might be more descriptive to assign `output_file_size + json_output_markdown_size <= max_allowed_bytes` to a variable that indicates whether `reportSizeWithinLimit`, which makes the other comments less necessary to understand what's going on

File: src/inputs.ts [hunk 5]
```diff
@@ -88,12 +151,28 @@ ${pretty_output_json}
   const should_append_json = pretty_output_json != '{}';
 
   if (output && output_text_description_file && should_append_json) {
```
Comment (ikatz-drizly) [inline src/inputs.ts:153]: This if/else block seems wrong.  It feels like `buildOutputDescription` function should be written with destructuring parameters (https://mariusschulz.com/blog/typing-destructured-object-parameters-in-typescript) and then within that function it would deal with the existence or non-existence of those values.  Then, it would be a single function call here instead of 3 variations.

File: src/inputs.ts (hunk 4 above)
Comment (tatchley-drizly) [reply src/inputs.ts:69]: > the else { } is unnecessary here.

Do you mean to just rework it into something like:
```
const reportSizeWithinLimit = output_file_size + json_output_markdown_size <= max_allowed_bytes
if (!reportSizeWithinLimit) {
  ...
  return corrected_content;
}
return output_description_md + json_vars; 
```
I use that function to handle all of the cases, so there would be times when nothing needs to be done and we can return the concatenation without any manipulation.

> it also might be more descriptive to assign ...  to a variable 
Happy to make this change

**Comment by ikatz-drizly** [comment]
Overall it looks like you've named things in `snake_case` where the convention here (and in [anecdotal Typescript docs](https://basarat.gitbook.io/typescript/styleguide)) seems to be `camelCase` for function and variable names.  @benreynolds-drizly or @mikekantzer-drizly may be able to chime in on that.

File: src/inputs.ts [hunk 6]
```diff
@@ -17,12 +41,51 @@ const parseJSON = <T>(getInput: GetInput, property: string): T | undefined => {
   }
 };
 
+function reverseString(input_string: string): string {
+  return input_string.split('').reverse().join('');
```
Comment (brettdugan-drizly) [inline src/inputs.ts:45]: Semi-irrelevant to utf-8 but this will not properly reverse some characters with surrogate pairs and it kind of promises that it will work on any string.

```suggestion
  return [...input_string].reverse().join('');
```

This change should deal with that problem better.

I'd also like to comment that we're mixing case and camel case in our TS. Camel case is the typical standard.

**Comment by ikatz-drizly** [comment]
We should do our own tests on whether the error message (chars vs bytes) is correct.  I'm sure there is a nice 4-byte sequence for some emoji we can abuse to test the theory.

File: src/inputs.ts (hunk 2 above)
Comment (brettdugan-drizly) [reply src/inputs.ts:19]: It should be able to be removed altogether if the parameter types are true. Because `function truncateToBinarySize(string: string, binaryMaxSize: number): string {` implies that  the paremeter`string` will always be of type `string`.

If you can't expect string to have a defined value it should be `string?: string` or `string = ''`

It's also good to avoid using type names as parameter or variable names in typescript. Something like `str: string` would be a little clearer.

File: src/inputs.ts (hunk 1 above)
Comment (brettdugan-drizly) [reply src/inputs.ts:16]: Ditto for 
`  const DEFAULT_TRUNCATE_STRING_BINARY_SIZE = Buffer.byteLength(DEFAULT_TRUNCATE_STRING, 'utf8');`

File: src/inputs.ts [hunk 7]
```diff
@@ -4,6 +4,30 @@ import * as fs from 'fs';
 
 type GetInput = (name: string, options?: InputOptions | undefined) => string;
 
+// getBinarySize and truncateToBinarySize were adapted from unlicensed work here:
+// https://github.com/lemonde/utf8-binary-cutter/blob/master/lib/utf8-binary-cutter.js
+function getBinarySize(s: string): number {
+  // Get binary size of a UTF8 string
+  return Buffer.byteLength(s || '', 'utf8');
```
Comment (brettdugan-drizly) [inline src/inputs.ts:11]: Same point about parameter types.

`s: string` implies that we expect `s` to always have a string value. If we thing this isn't true at run time then it should be `s?: string` or `s = ''` if we want a default.

File: src/inputs.ts (hunk 5 above)
Comment (brettdugan-drizly) [reply src/inputs.ts:153]: I think we should try to decompose this function in general. It's a little hard to follow as is. A lot of error handling in different places and conditions that could be moved elsewhere. For example, I wouldn't expect a function called `parseInputs` to write a file.

**Comment by brettdugan-drizly** [comment]
> Overall it looks like you've named things in `snake_case` where the convention here (and in [anecdotal Typescript docs](https://basarat.gitbook.io/typescript/styleguide)) seems to be `camelCase` for function and variable names. @benreynolds-drizly or @mikekantzer-drizly may be able to chime in on that.

should be camel

**Comment by brettdugan-drizly** [comment]
> We should do our own tests on whether the error message (chars vs bytes) is correct. I'm sure there is a nice 4-byte sequence for some emoji we can abuse to test the theory.

Agree. Let me know if you need help setting up testing @tatchley-drizly

**Comment by tatchley-drizly** [comment]
> > We should do our own tests on whether the error message (chars vs bytes) is correct. I'm sure there is a nice 4-byte sequence for some emoji we can abuse to test the theory.
> 
> Agree. Let me know if you need help setting up testing @tatchley-drizly

@ikatz-drizly @brettdugan-drizly After [poking around with the API](https://github.com/DrizlyInc/character-test/runs/3403287547?check_suite_focus=true), I found out the following things:
* The 65535 limit is indeed a limit on the number of _characters_, not the number of bytes.
* This limit applies independently to both the `output.summary` and `output.text` fields

I'll start updating this PR to reflect character length as the limiting factor (as well as the previous suggestions).

**Comment by tatchley-drizly** [comment]
We have decided to move [unit-testing to another ticket](https://drizly.atlassian.net/browse/DVX-597) in the interest of unblocking further integrations.
